pymc/step_methods/compound.py

```diff
@@ -48,9 +50,61 @@ class Competence(IntEnum):
     IDEAL = 3
 
 
+def infer_warn_stats_info(
+    stats_dtypes: List[Dict[str, StatDtype]],
+    sds: Dict[str, Tuple[StatDtype, StatShape]],
+    stepname: str,
+) -> Tuple[List[Dict[str, StatDtype]], Dict[str, Tuple[StatDtype, StatShape]]]:
+    """Helper function to get `stats_dtypes` and `stats_dtypes_shapes` from either of them."""
+    # Avoid side-effects on the original lists/dicts
+    stats_dtypes = [d.copy() for d in stats_dtypes]
+    sds = sds.copy()
+    # Disallow specification of both attributes
+    if stats_dtypes and sds:
+        raise TypeError(
+            "Only one of `stats_dtypes_shapes` or `stats_dtypes` must be specified."
+            f" `{stepname}.stats_dtypes` should be removed."
+        )
+
+    # Infer one from the other
+    if not sds and stats_dtypes:
+        warnings.warn(
+            f"`{stepname}.stats_dtypes` is deprecated."
+            " Please update it to specify `stats_dtypes_shapes` instead.",
+            DeprecationWarning,
+        )
+        if len(stats_dtypes) > 1:
+            raise TypeError(
+                f"`{stepname}.stats_dtypes` must be a list containing at most one dict."
+            )
+        for sd in stats_dtypes:
+            for sname, dtype in sd.items():
+                sds[sname] = (dtype, None)
+    elif sds:
+        stats_dtypes.append({sname: dtype for sname, (dtype, _) in sds.items()})
+    return stats_dtypes, sds
+
+
 class BlockedStep(ABC):
     stats_dtypes: List[Dict[str, type]] = []
+    """A list containing <=1 dictionary that maps stat names to dtypes.
+
+    This attribute is deprecated.
+    Use `stats_dtypes_shapes` instead.
+    """
+
+    stats_dtypes_shapes: Dict[str, Tuple[StatDtype, StatShape]] = {}
+    """Maps stat names to dtypes and shapes.
+
+    Shapes are interpreted in the following ways:
+    - `[]` is a scalar.
+    - `[3,]` is a length-3 vector.
+    - `[4, 0]` is a matrix with 4 rows and a dynamic number of columns.
```

Comment:
It seems that both `0` and `None` can express that the stat shape is dynamic (either one dimension of it, or all/unknown dimensions). Is there a reason to prefer `0` for the first case and `None` for the second case?

One consideration is that shape tuples including 0 mean something different in numpy than what we are meaning here. I.e. `np.ones((..., 0, ...))` will always be empty even if the ellipses are filled with positive values. Perhaps it could be more clear to prefer `None` for both cases? So `[4, None]` is a matrix with 4 rows and a dynamic number of columns.

Reply:
Although this suggestion overloads the use of `None` when specifying shape info to pytensor, it does so in the same spirit as pytensor's behavior

Reply:
This is mostly motivated by the way how McBackend encodes shapes into protocol buffers, which is by a `repeated uint64`. I discussed this in chat with @ricardoV94 who proposed to use `-1` for dynamic dimensions and have `0` with the same interpretation as NumPy like you wrote.

Reply:
Yeah, we should use None and then convert that to 0 or -1 in the backend. Or must we do the conversion ourselves here?

Reply:
Conversion must be done before the shape is passed to `mcb.Variable` which is going to be located in `pm.backends.mcbackend`

Reply:
I changed the docstring to specify `-1` for dynamic dims as discussed